[Bug] Hive read external failed when table schema is not equal data schema and filter push down actively

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.


### Paimon version

paimon-hive-connector-3.1-0.7-SNAPSHOT.jar

### Compute Engine

hive3.1.3

### Minimal reproduce step

0. use sparksql to creata table with uppercase field and use lowercase field query by hivesql.
1. create a paimon table and insert some data.(Make sure has some data file use older schema file)
2. [SparkSQL]do some ddl which will generate a new version `schema-n` file
3. [HiveSQL] select * from tbl where col is not null.(Make sure has filter push down)
4. Catch NPE Log
`Error: java.io.IOException: java.lang.NullPointerException: Find no field ecsource
at org.apache.hadoop.hive.io.HiveIOExceptionHandlerChain.handleRecordReaderCreationException(HiveIOExceptionHandlerChain.java:97)
at org.apache.hadoop.hive.io.HiveIOExceptionHandlerUtil.handleRecordReaderCreationException(HiveIOExceptionHandlerUtil.java:57)
at org.apache.hadoop.hive.ql.io.HiveInputFormat.getRecordReader(HiveInputFormat.java:420)
at org.apache.hadoop.hive.ql.io.CombineHiveInputFormat.getRecordReader(CombineHiveInputFormat.java:704)
at org.apache.hadoop.mapred.MapTask$TrackedRecordReader.<init>(MapTask.java:175)
at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:444)
at org.apache.hadoop.mapred.MapTask.run(MapTask.java:349)
at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:174)
at java.security.AccessController.doPrivileged(Native Method)
at javax.security.auth.Subject.doAs(Subject.java:422)
at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1730)
at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:168)
Caused by: java.lang.NullPointerException: Find no field ecsource
at org.apache.paimon.utils.Preconditions.checkNotNull(Preconditions.java:65)
at org.apache.paimon.schema.SchemaEvolutionUtil.lambda$createDataFilters$7(SchemaEvolutionUtil.java:300)
at org.apache.paimon.predicate.LeafPredicate.visit(LeafPredicate.java:124)
at org.apache.paimon.predicate.PredicateReplaceVisitor.visit(PredicateReplaceVisitor.java:32)
at org.apache.paimon.predicate.PredicateReplaceVisitor.visit(PredicateReplaceVisitor.java:26)
at org.apache.paimon.predicate.CompoundPredicate.visit(CompoundPredicate.java:73)
at org.apache.paimon.schema.SchemaEvolutionUtil.createDataFilters(SchemaEvolutionUtil.java:334) at org.apache.paimon.utils.BulkFormatMapping$BulkFormatMappingBuilder.build(BulkFormatMapping.java:164)
at org.apache.paimon.io.KeyValueFileReaderFactory.lambda$createRecordReader$1(KeyValueFileReaderFactory.java:105) at org.apache.paimon.io.KeyValueFileReaderFactory.lambda$createRecordReader$2(KeyValueFileReaderFactory.java:114)
at java.util.HashMap.computeIfAbsent(HashMap.java:1128) at org.apache.paimon.io.KeyValueFileReaderFactory.createRecordReader(KeyValueFileReaderFactory.java:112)
at org.apache.paimon.io.KeyValueFileReaderFactory.createRecordReader(KeyValueFileReaderFactory.java:91)
at org.apache.paimon.mergetree.MergeTreeReaders.lambda$readerForRun$2(MergeTreeReaders.java:88)
at org.apache.paimon.mergetree.compact.ConcatRecordReader.create(ConcatRecordReader.java:50)
at org.apache.paimon.mergetree.MergeTreeReaders.readerForRun(MergeTreeReaders.java:94)
at org.apache.paimon.mergetree.MergeTreeReaders.lambda$readerForSection$1(MergeTreeReaders.java:77)
at org.apache.paimon.mergetree.MergeSorter.mergeSort(MergeSorter.java:113)
at org.apache.paimon.mergetree.MergeTreeReaders.readerForSection(MergeTreeReaders.java:79)
at org.apache.paimon.operation.KeyValueFileStoreRead.lambda$batchMergeRead$4(KeyValueFileStoreRead.java:252)
at org.apache.paimon.mergetree.compact.ConcatRecordReader.create(ConcatRecordReader.java:50)
at org.apache.paimon.operation.KeyValueFileStoreRead.batchMergeRead(KeyValueFileStoreRead.java:262)
at org.apache.paimon.operation.KeyValueFileStoreRead.createReaderWithoutOuterProjection(KeyValueFileStoreRead.java:222) at org.apache.paimon.operation.KeyValueFileStoreRead.createReader(KeyValueFileStoreRead.java:199)
at org.apache.paimon.table.source.KeyValueTableRead.reader(KeyValueTableRead.java:53)
at org.apache.paimon.table.source.AbstractDataTableRead.createReader(AbstractDataTableRead.java:84) at org.apache.paimon.hive.mapred.PaimonRecordReader.<init>(PaimonRecordReader.java:99)
at org.apache.paimon.hive.mapred.PaimonInputFormat.getRecordReader(PaimonInputFormat.java:159)
at org.apache.hadoop.hive.ql.io.HiveInputFormat.getRecordReader(HiveInputFormat.java:417)`

it's seem that because of tableschema.id != dataschema.id, so that will call `SchemaEvolutionUtil.createDataFilters(tableSchema.fields(), dataSchema.fields(), filters)`. However, the field name in filters is lowercase because of we use hive query engine, by it is uppercase in tableSchema. 

![image](https://github.com/apache/incubator-paimon/assets/31072303/b7bc595b-85ce-4acf-bcc4-0d3850754930)
![image](https://github.com/apache/incubator-paimon/assets/31072303/9aa35974-5ac3-44ef-88b4-0d46000abf42)




### What doesn't meet your expectations?

hivesql query paimon external table can work after schema evolution. whatever use uppercase or lowercase field name in sql.

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/paimon-hive/paimon-hive-connector-common/src/main/java/org/apache/paimon/hive/SearchArgumentToPredicateConverter.java
+++ b/paimon-hive/paimon-hive-connector-common/src/main/java/org/apache/paimon/hive/SearchArgumentToPredicateConverter.java
@@ -49,7 +49,7 @@ public class SearchArgumentToPredicateConverter {
 
     private final ExpressionTree root;
     private final List<PredicateLeaf> leaves;
-    private final List<String> columnNames;
+    private final List<String> hiveColumnNames;
     private final List<DataType> columnTypes;
     @Nullable private final Set<String> readColumnNames;
     private final PredicateBuilder builder;
@@ -61,7 +61,7 @@ public class SearchArgumentToPredicateConverter {
             @Nullable Set<String> readColumnNames) {
         this.root = searchArgument.getExpression();
         this.leaves = searchArgument.getLeaves();
-        this.columnNames =
+        this.hiveColumnNames =
                 columnNames.stream().map(String::toLowerCase).collect(Collectors.toList());
         this.columnTypes = columnTypes;
         if (readColumnNames != null) {
@@ -74,7 +74,7 @@ public class SearchArgumentToPredicateConverter {
                 new PredicateBuilder(
                         RowType.of(
                                 this.columnTypes.toArray(new DataType[0]),
-                                this.columnNames.toArray(new String[0])));
+                                columnNames.toArray(new String[0])));
     }
 
     public Optional<Predicate> convert() {
@@ -140,7 +140,7 @@ public class SearchArgumentToPredicateConverter {
                             + " is a partition column.");
         }
 
-        int idx = columnNames.indexOf(columnName);
+        int idx = hiveColumnNames.indexOf(columnName);
         Preconditions.checkArgument(idx >= 0, "Column " + columnName + " not found.");
         DataType columnType = columnTypes.get(idx);
         switch (leaf.getOperator()) {

--- a/paimon-hive/paimon-hive-connector-common/src/test/java/org/apache/paimon/hive/HiveReadITCase.java
+++ b/paimon-hive/paimon-hive-connector-common/src/test/java/org/apache/paimon/hive/HiveReadITCase.java
@@ -73,8 +73,7 @@ public class HiveReadITCase extends HiveTestBase {
                         "");
         Identifier identifier = Identifier.create(DATABASE_TEST, tableName);
         Path tablePath = AbstractCatalog.newTableLocation(path, identifier);
-        SchemaManager schemaManager = new SchemaManager(LocalFileIO.create(), tablePath);
-        schemaManager.createTable(schema);
+        new SchemaManager(LocalFileIO.create(), tablePath).createTable(schema);
 
         // Create hive external table
         String hiveSql =
@@ -85,7 +84,6 @@ public class HiveReadITCase extends HiveTestBase {
                                 "STORED BY '" + PaimonStorageHandler.class.getName() + "'",
                                 "LOCATION '" + tablePath.toUri().toString() + "'"));
         assertThatCode(() -> hiveShell.execute(hiveSql)).doesNotThrowAnyException();
-
         List<String> result = hiveShell.executeQuery("SHOW CREATE TABLE " + tableName);
         assertThat(result)
                 .containsAnyOf(
@@ -98,32 +96,18 @@ public class HiveReadITCase extends HiveTestBase {
                         "  'org.apache.paimon.hive.PaimonStorageHandler' ");
 
         hiveShell.execute("INSERT INTO " + tableName + " VALUES (1,'Hello'),(2,'Paimon')");
-        hiveShell.execute("Drop Table " + tableName);
         result = hiveShell.executeQuery("SELECT col2, col1 FROM " + tableName);
         assertThat(result).containsExactly("Hello\t1", "Paimon\t2");
         result = hiveShell.executeQuery("SELECT col2 FROM " + tableName);
         assertThat(result).containsExactly("Hello", "Paimon");
         result = hiveShell.executeQuery("SELECT Col2 FROM " + tableName);
         assertThat(result).containsExactly("Hello", "Paimon");
-
         result = hiveShell.executeQuery("SELECT * FROM " + tableName + " WHERE col2 = 'Hello'");
         assertThat(result).containsExactly("1\tHello");
         result =
                 hiveShell.executeQuery(
                         "SELECT * FROM " + tableName + " WHERE Col2 in ('Hello', 'Paimon')");
         assertThat(result).containsExactly("1\tHello", "2\tPaimon");
-
-        schemaManager.commitChanges(SchemaChange.addColumn("D1", DataTypes.STRING()));
-        String hiveSql1 =
-                String.join(
-                        "\n",
-                        Arrays.asList(
-                                "CREATE EXTERNAL TABLE " + tableName + " ",
-                                "STORED BY '" + PaimonStorageHandler.class.getName() + "'",
-                                "LOCATION '" + tablePath.toUri().toString() + "'"));
-        assertThatCode(() -> hiveShell.execute(hiveSql1)).doesNotThrowAnyException();
-        result = hiveShell.executeQuery("SELECT * FROM " + tableName + " WHERE Col2 is not null");
-        System.out.println(result);
     }
 
     @Test

--- a/paimon-hive/paimon-hive-connector-common/src/test/java/org/apache/paimon/hive/HiveReadITCase.java
+++ b/paimon-hive/paimon-hive-connector-common/src/test/java/org/apache/paimon/hive/HiveReadITCase.java
@@ -18,16 +18,28 @@
 
 package org.apache.paimon.hive;
 
+import org.apache.paimon.CoreOptions;
 import org.apache.paimon.catalog.AbstractCatalog;
 import org.apache.paimon.catalog.Identifier;
+import org.apache.paimon.data.BinaryString;
+import org.apache.paimon.data.GenericRow;
+import org.apache.paimon.data.InternalRow;
 import org.apache.paimon.fs.Path;
 import org.apache.paimon.fs.local.LocalFileIO;
 import org.apache.paimon.hive.mapred.PaimonInputFormat;
 import org.apache.paimon.hive.mapred.PaimonRecordReader;
+import org.apache.paimon.options.CatalogOptions;
+import org.apache.paimon.options.Options;
 import org.apache.paimon.schema.Schema;
+import org.apache.paimon.schema.SchemaChange;
 import org.apache.paimon.schema.SchemaManager;
+import org.apache.paimon.table.Table;
+import org.apache.paimon.table.sink.StreamTableCommit;
+import org.apache.paimon.table.sink.StreamTableWrite;
+import org.apache.paimon.table.sink.StreamWriteBuilder;
 import org.apache.paimon.types.DataField;
 import org.apache.paimon.types.DataTypes;
+import org.apache.paimon.types.RowType;
 
 import org.apache.paimon.shade.guava30.com.google.common.collect.Lists;
 import org.apache.paimon.shade.guava30.com.google.common.collect.Maps;
@@ -45,7 +57,7 @@ import static org.assertj.core.api.Assertions.assertThatCode;
 public class HiveReadITCase extends HiveTestBase {
 
     @Test
-    public void testReExternalTableWithIgnoreCase() throws Exception {
+    public void testReadExternalTableWithEmptyDataAndIgnoreCase() throws Exception {
         // Create hive external table with paimon table
         String tableName = "with_ignore_case";
 
@@ -61,7 +73,8 @@ public class HiveReadITCase extends HiveTestBase {
                         "");
         Identifier identifier = Identifier.create(DATABASE_TEST, tableName);
         Path tablePath = AbstractCatalog.newTableLocation(path, identifier);
-        new SchemaManager(LocalFileIO.create(), tablePath).createTable(schema);
+        SchemaManager schemaManager = new SchemaManager(LocalFileIO.create(), tablePath);
+        schemaManager.createTable(schema);
 
         // Create hive external table
         String hiveSql =
@@ -85,6 +98,7 @@ public class HiveReadITCase extends HiveTestBase {
                         "  'org.apache.paimon.hive.PaimonStorageHandler' ");
 
         hiveShell.execute("INSERT INTO " + tableName + " VALUES (1,'Hello'),(2,'Paimon')");
+        hiveShell.execute("Drop Table " + tableName);
         result = hiveShell.executeQuery("SELECT col2, col1 FROM " + tableName);
         assertThat(result).containsExactly("Hello\t1", "Paimon\t2");
         result = hiveShell.executeQuery("SELECT col2 FROM " + tableName);
@@ -98,5 +112,77 @@ public class HiveReadITCase extends HiveTestBase {
                 hiveShell.executeQuery(
                         "SELECT * FROM " + tableName + " WHERE Col2 in ('Hello', 'Paimon')");
         assertThat(result).containsExactly("1\tHello", "2\tPaimon");
+
+        schemaManager.commitChanges(SchemaChange.addColumn("D1", DataTypes.STRING()));
+        String hiveSql1 =
+                String.join(
+                        "\n",
+                        Arrays.asList(
+                                "CREATE EXTERNAL TABLE " + tableName + " ",
+                                "STORED BY '" + PaimonStorageHandler.class.getName() + "'",
+                                "LOCATION '" + tablePath.toUri().toString() + "'"));
+        assertThatCode(() -> hiveShell.execute(hiveSql1)).doesNotThrowAnyException();
+        result = hiveShell.executeQuery("SELECT * FROM " + tableName + " WHERE Col2 is not null");
+        System.out.println(result);
+    }
+
+    @Test
+    public void testReadExternalTableWithDataAndIgnoreCase() throws Exception {
+        // Create hive external table with paimon table
+        String tableName = "with_data_and_ignore_case";
+
+        // Create a paimon table
+        Identifier identifier = Identifier.create(DATABASE_TEST, tableName);
+
+        Options conf = new Options();
+        conf.set(CatalogOptions.WAREHOUSE, path);
+        conf.set(CoreOptions.BUCKET, 2);
+        conf.set(CoreOptions.FILE_FORMAT, CoreOptions.FileFormatType.AVRO);
+        RowType.Builder rowType = RowType.builder();
+        rowType.field("col1", DataTypes.INT());
+        rowType.field("Col2", DataTypes.STRING());
+
+        Table table =
+                FileStoreTestUtils.createFileStoreTable(
+                        conf,
+                        rowType.build(),
+                        Collections.emptyList(),
+                        Collections.emptyList(),
+                        identifier);
+
+        // insert data into paimon table, make sure has some data file use older schema file.
+        List<InternalRow> data =
+                Arrays.asList(
+                        GenericRow.of(1, BinaryString.fromString("Hello")),
+                        GenericRow.of(2, BinaryString.fromString("Paimon")));
+
+        StreamWriteBuilder streamWriteBuilder = table.newStreamWriteBuilder();
+        StreamTableWrite write = streamWriteBuilder.newWrite();
+        StreamTableCommit commit = streamWriteBuilder.newCommit();
+        for (InternalRow rowData : data) {
+            write.write(rowData);
+        }
+        commit.commit(0, write.prepareCommit(true, 0));
+        write.close();
+        commit.close();
+
+        // add column, do some ddl which will generate a new version schema-n file.
+        Path tablePath = AbstractCatalog.newTableLocation(path, identifier);
+        SchemaManager schemaManager = new SchemaManager(LocalFileIO.create(), tablePath);
+        schemaManager.commitChanges(SchemaChange.addColumn("N1", DataTypes.STRING()));
+
+        // Create hive external table
+        String hiveSql =
+                String.join(
+                        "\n",
+                        Arrays.asList(
+                                "CREATE EXTERNAL TABLE " + tableName + " ",
+                                "STORED BY '" + PaimonStorageHandler.class.getName() + "'",
+                                "LOCATION '" + tablePath.toUri().toString() + "'"));
+        assertThatCode(() -> hiveShell.execute(hiveSql)).doesNotThrowAnyException();
+
+        List<String> result =
+                hiveShell.executeQuery("SELECT * FROM " + tableName + " WHERE col2 is not null");
+        assertThat(result).containsExactly("1\tHello\tNULL", "2\tPaimon\tNULL");
     }
 }

